runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/util/BoundedQueueExecutorTest.java

```diff
@@ -0,0 +1,232 @@
+
+    // Max pool size was reached so no new work is accepted.
+    executor.execute(m3, 1);
+    assertFalse(processStart3.await(1000, TimeUnit.MILLISECONDS));
+
+    // Increase the max thread count
+    executor.setMaximumPoolSize(3, 103);
+    assertEquals(3, executor.maximumThreadCount());
+
+    // m3 is accepted
+    processStart3.await();
+    assertEquals(3, executor.activeCount());
+
+    stop.countDown();
+    executor.shutdown();
+  }
+
+  @Test
+  public void testRecordTotalTimeMaxActiveThreadsUsed() throws Exception {
+    CountDownLatch processStart1 = new CountDownLatch(1);
+    CountDownLatch processStart2 = new CountDownLatch(1);
+    CountDownLatch processStart3 = new CountDownLatch(1);
+    CountDownLatch stop = new CountDownLatch(1);
+    Runnable m1 = createSleepProcessWorkFn(processStart1, stop);
+    Runnable m2 = createSleepProcessWorkFn(processStart2, stop);
+    Runnable m3 = createSleepProcessWorkFn(processStart3, stop);
+
+    // Initial state.
+    assertEquals(0, executor.activeCount());
+    assertEquals(2, executor.maximumThreadCount());
+
+    // m1 and m2 are accepted.
+    executor.execute(m1, 1);
+    processStart1.await();
+    assertEquals(1, executor.activeCount());
+    executor.execute(m2, 1);
+    processStart2.await();
+    assertEquals(2, executor.activeCount());
+
+    // Max pool size was reached so no new work is accepted.
+    executor.execute(m3, 1);
+    assertFalse(processStart3.await(1000, TimeUnit.MILLISECONDS));
+
+    assertEquals(0l, executor.allThreadsActiveTime());
+    stop.countDown();
+    while (executor.activeCount() != 0) {
+      // Waiting for all threads to be ended.
+    }
+    // Max pool size was reached so the allThreadsActiveTime() was updated.
+    assertThat(executor.allThreadsActiveTime(), greaterThan(0l));
+
+    executor.shutdown();
+  }
+
+  @Test
+  public void testRecordTotalTimeMaxActiveThreadsUsedWhenMaximumThreadCountUpdated()
+      throws Exception {
+    CountDownLatch processStart1 = new CountDownLatch(1);
+    CountDownLatch processStart2 = new CountDownLatch(1);
+    CountDownLatch processStart3 = new CountDownLatch(1);
+    CountDownLatch stop = new CountDownLatch(1);
+    Runnable m1 = createSleepProcessWorkFn(processStart1, stop);
+    Runnable m2 = createSleepProcessWorkFn(processStart2, stop);
+    Runnable m3 = createSleepProcessWorkFn(processStart3, stop);
+
+    // Initial state.
+    assertEquals(0, executor.activeCount());
+    assertEquals(2, executor.maximumThreadCount());
+
+    // m1 and m2 are accepted.
+    executor.execute(m1, 1);
+    processStart1.await();
+    assertEquals(1, executor.activeCount());
+    executor.execute(m2, 1);
+    processStart2.await();
+    assertEquals(2, executor.activeCount());
+
+    // Max pool size was reached so no new work is accepted.
+    executor.execute(m3, 1);
+    assertFalse(processStart3.await(1000, TimeUnit.MILLISECONDS));
+
+    assertEquals(0l, executor.allThreadsActiveTime());
```

Comment:
this made me realize we're only updating the count after the active work completes.  That could delay possibly requesting additional threads if there is long-running work on the available threads.  Maybe in a followup we might want to periodically increase this with some monitoring thread instead of relying on the finishing threads themselves to increment.

Reply:
Yeah if all the threads are occupied by the long-running tasks, we may under-estimate the related time counters. Not sure if we use these counters for any decisions now but if later we decide to tune the thread count based on these time counters and thread utilization, it might cause problem. Let me create a bug to follow up on this.